apps/src/blockly/addons/cdoUtils.js

```diff
@@ -14,27 +14,143 @@ import {
+  // TODO: When we add behaviors, we should always hide behavior blocks.
+  const procedureTypesToHide = [];
+  if (
+    Blockly.useModalFunctionEditor &&
+    experiments.isEnabled(experiments.MODAL_FUNCTION_EDITOR)
+  ) {
+    procedureTypesToHide.push('procedures_defnoreturn');
+  }
+  moveHiddenProcedures(
+    parsedSource,
+    parsedHiddenDefinitions,
+    procedureTypesToHide
+  );
+  return {parsedSource, parsedHiddenDefinitions, blockOrderMap};
+}
+
+/**
+ * Convert source and hidden definitions to parsed json objects. If source was xml,
+ * convert to json before parsing. Also create a block order map if source was xml, which will
+ * allow us to correctly place blocks on the workspace.
+ * @param {string} source - workspace serialization, either XML or JSON
+ * @param {[string]} hiddenDefinitions - hidden definitions serialization, in JSON. Only used in Google Blockly labs.
+ * @returns {parsedSource: Object, parsedHiddenDefinitions: Object, blockOrderMap: Object}
+ */
+function parseSourceAndHiddenDefinitions(source, hiddenDefinitions) {
   let isXml = stringIsXml(source);
-  let stateToLoad;
+  let parsedSource;
+  let parsedHiddenDefinitions;
   let blockOrderMap;
+  hiddenDefinitions = hiddenDefinitions || '{}';
   if (isXml) {
     const xml = parseXmlElement(source);
-    stateToLoad = convertXmlToJson(xml);
+    parsedSource = convertXmlToJson(xml);
     blockOrderMap = Blockly.Xml.createBlockOrderMap(xml);
   } else {
-    stateToLoad = JSON.parse(source);
+    parsedSource = JSON.parse(source);
+  }
+  parsedHiddenDefinitions = JSON.parse(hiddenDefinitions);
+  return {parsedSource, parsedHiddenDefinitions, blockOrderMap};
+}
+
+/**
+ * Move hidden procedures from the source to the hidden definition object.
+ * These will be used to initialize the main and hidden definitions workspaces, respectively.
+ * Procedures are hidden if they have a type in the procedureTypesToHide array.
+ * In addition, copy the procedure model from the source
+ * object to the hidden definition object when moving a procedure.
+ * @param {Object} source Project source object, parsed from JSON.
+ * @param {Object} hiddenDefinitions Hidden Definition object, parsed from JSON (or an empty object)
+ * @param {Array<string>} procedureTypesToHide procedure types to move to procedures object.
+ * @returns void
+ */
+function moveHiddenProcedures(source, hiddenDefinitions, procedureTypesToHide) {
+  if (
+    procedureTypesToHide.length === 0 ||
+    !source.blocks ||
+    !source.blocks.blocks
+  ) {
+    return;
+  }
+  const blocksToHide = [];
+  const otherBlocks = [];
+  source.blocks.blocks.forEach(block => {
+    if (procedureTypesToHide.includes(block.type)) {
+      blocksToHide.push(block);
+    } else {
+      otherBlocks.push(block);
+    }
+  });
+  source.blocks.blocks = otherBlocks;
+  hiddenDefinitions.blocks ||= {};
+  hiddenDefinitions.blocks.blocks ||= [];
+  hiddenDefinitions.blocks.blocks.push(...blocksToHide);
```

Comment:
Honestly I just don't use lodash much so I forget about it. I don't think there's any issue with using it here, I can update it since that seems cleaner!

Reply:
we don't want to a pure set here since if there are existing blocks in the hidden definition workspace we want to keep them. Is there a lodash function that is useful in that case?

Reply:
You could `get` and then `set` if that still feels easier?
```
const existingBlocks = get(hiddenDefinitions, 'blocks.blocks', []);
set(hiddenDefinitions, 'blocks.blocks', [...existingBlocks, ...blocksToHide]);
```

Reply:
Existing documentation:
- https://lodash.com/docs/4.17.15#get
- https://lodash.com/docs/4.17.15#set

Noteworthy for `set`: "If a portion of the path doesn't exist, it's created. Arrays are created for missing index properties while objects are created for all other missing properties." That sounds like what we want.

Reply:
updated!